fuse/dl/losses/contrastive/contrastive_loss.py

```diff
@@ -0,0 +1,112 @@
+from typing import Callable, Optional
+
+from fuse.dl.losses.loss_base import LossBase
+from fuse.utils.ndict import NDict
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor
+
+
+def softcrossentropyloss(target: Tensor, logits: Tensor) -> Tensor:
+    """
+    From the pytorch discussion Forum:
+    https://discuss.pytorch.org/t/soft-cross-entropy-loss-tf-has-it-does-pytorch-have-it/69501
+    """
+    logprobs = F.log_softmax(logits, dim=1)
+    # print(target * logprobs)
+    loss = -(target * logprobs).sum() / target.sum()  # logits.shape[0]
+    return loss
+
+
+def supervised_contrastive_loss(
+    feat_xi: Tensor, feat_xj: Tensor, yi: Tensor, yj: Tensor, symm: bool = False, alpha: float = 0.5, temp: float = 0.2
+) -> Tensor:
+    if feat_xi.isnan().any() | feat_xj.isnan().any():
+        print("feature has nan")
+    if len(feat_xi.shape) < 2:
+        feat_xi = feat_xi.unsqueeze(dim=0)
+    if len(feat_xj.shape) < 2:
+        feat_xj = feat_xj.unsqueeze(dim=0)
+    feat_xi = F.normalize(feat_xi, p=2, dim=1)
+    feat_xj = F.normalize(feat_xj, p=2, dim=1)
+    label_vec = torch.unsqueeze(yi, 0)
+    mask = torch.eq(torch.transpose(label_vec, 0, 1), yj).float()
+    if mask.sum() == 0:
+        return torch.tensor(0.0).to("cuda")
+    logits_xi = torch.div(torch.matmul(feat_xi, torch.transpose(feat_xj, 0, 1)), temp)
+    logits_max, _ = torch.max(logits_xi, dim=1, keepdim=True)
+    logits_xi = logits_xi - logits_max.detach()
+    loss_xi = softcrossentropyloss(mask, logits_xi)  # / torch.sum(mask, 1)
+    if symm:
+        logits_xj = torch.matmul(feat_xj, torch.transpose(feat_xi, 0, 1)) / temp
+        loss_xj = softcrossentropyloss(mask, logits_xj) / torch.sum(mask, 0)
+        return alpha * loss_xi.sum() + (1 - alpha) * loss_xj.sum()
+    else:
+        return loss_xi.mean()
+
+
+class SupervisedContrastiveLoss(LossBase):
+    """
+    Default Fuse loss function
+
+    '''
+
+    """
+
+    def __init__(
+        self,
+        *,  # prevent positional args
+        feat_i: str = None,
+        feat_j: str = None,
+        target_i: str = None,
+        target_j: str = None,
+        temp: float = 0.2,
+        alpha: float = 0.5,
+        symmetrical: bool = False,
+        callable: Callable = None,
+        weight: Optional[float] = None,
+        preprocess_func: Optional[Callable] = None,
+    ) -> None:
+        """
+        This class wraps a PyTorch loss function with a Fuse api.
```

Comment:
Update the comments here - be specific about what is expected to be in batch_dict[feat_i], batch_dict[targe_i], ...